### ISSUE TITLE
Feature Request: support mhtml

### ISSUE THREAD

**Issue by monotseng** [nuthrash/obsidian-html-plugin#19]
Hi

Could you also support files with the extension .mhtml? Because when saving some web pages, .mhtml files can represent the original web page style to the greatest extent with only one file, compared to .html files. 

Thank you for your hard work.

**Comment by nuthrash** [comment]
The .mhtml file format is not a standard HTML format, and it seems that there is no document or book has description about its format. I have tried to investigate any Javascript/Typescript library can parse and render it, but there is no workable library can integrate with Obsidian.
If you have any workable, free and open source library can provide to me, I might accomplish your feature request.

The alternative solution is to re-save the .mhtml files by the SingleFile web browser extension as new .html files.

**Comment by Marvin-Brouwer** [comment]
Would you be open to faking it by showing it inside of a page that only contains an iframe to the file?
If it's all local that should work I think?
Maybe there's some strange edge case nobody knows about, but, that can be POC'd.

**Comment by nuthrash** [comment]
> Would you be open to faking it by showing it inside of a page that only contains an iframe to the file? If it's all local that should work I think? Maybe there's some strange edge case nobody knows about, but, that can be POC'd.

Your suggestion is not work.
From https://issues.chromium.org/issues/41133520#comment6, it comments 
>For security reasons, MHTML files should only be loaded in top-frames.

Of course, I had tested this approach with Chrome, Firefox, and Obsidian.

**Comment by Marvin-Brouwer** [comment]
Yeah, I was hoping maybe because it's not a browser we can get away with it.
I did a littlebit of digging myself. According to WikiPedia: "The .mhtml and [.eml](https://en.wikipedia.org/wiki/Email#Filename_extensions) filename extensions are interchangeable".
And there seems to be an EML spec: https://docs.oasis-open.org/election/eml/v7.0/eml-v7.0.html.
I truly understand if you're not feeling like implementing this, I just figured I'd share this.

**Comment by gildas-lormeau** [comment]
Hi, this might interest you: https://github.com/gildas-lormeau/mhtml-to-html

**Comment by nuthrash** [comment]
> Hi, this might interest you: https://github.com/gildas-lormeau/mhtml-to-html

Oh, it looks good! I will try to integrate it with this plugin.
Thank you very much.

**Comment by nuthrash** [comment]
Hi @monotseng  and @Marvin-Brouwer.
I think this feature request can be accomplished by integrating with the mhtml-to-html library.
Below is the screenshot:

![Image](https://github.com/user-attachments/assets/34581d15-a29e-4586-b4b2-6e9b9dff87a8)

And I will add a new setting for this FR, the "MHTML File Format Supported".

![Image](https://github.com/user-attachments/assets/ae1ec753-82fd-484a-9f48-cbb86613c776)


There are some other FRs code need testing, so please be patient to wait for next version of this plugin. 

---

Thanks @gildas-lormeau again, your great projects make these features accomplishable.

**Comment by Marvin-Brouwer** [comment]
Awesome! No rush, I'm happy you're willing to add this.
If you like a second pair of eyes, I can offer a quick code review on a PR or something.

**Comment by nuthrash** [comment]
> If you like a second pair of eyes, I can offer a quick code review on a PR or something.

First of all, thank you.
In fact, there are few codes need to be reviewed (all heavy duty works are took by the mhtml-to-html library), and I am collecting different MHTML files and hyperlinks (other FPs' approaches) to test to see if I miss something.
If you provide or tell me where are public MHTML files could accessible, it would be very helpful.

**Comment by Marvin-Brouwer** [comment]
You can hit CTRL+S on any wepage and save it.
Make sure you select mhtml though otherwise you get a folder full of random files.
(I know for sure this works in chrome, I can imagine any browser can do it though)

**Comment by gildas-lormeau** [comment]
@nuthrash You can also collect some MHTML files on github with this search request: https://github.com/search?q=path%3A*.mht*&type=code

**Comment by nuthrash** [comment]
Hi @monotseng and @Marvin-Brouwer.

I have released version 1.0.13 to accomplish this feature request.
You might update to this version and try to open some MHTML files to see if there is any problem.

**Comment by Marvin-Brouwer** [comment]
@nuthrash looking good, I use this to save webpages when I ask an LLM something and they don't support exporting the chat.
So far chatgpt and lumo pages look fine!

Minor comments I have:
I see "No file" for a couple of seconds when opening an mhtml file. I imagine that's because it's still being generated. And, obsidian can't find the file yet. 
Perhaps you can start with a default HTML page that shows some kind of loading text? 

Additionally, I have a couple of very big mhtml files which seem to take quite a while to reopen.
I'm not sure if it's because of rendering or the on the fly HTML conversion. But perhaps some caching mechanism may help here?

Having said all that, I'm happy to create separate github issues with low prio for both of those.
I'm really happy with this feature as it is.

Thanks!

**Comment by gildas-lormeau** [comment]
@Marvin-Brouwer, the perf issue is probably related to `mtml-to-html`. If you can reproduce it with [this demo](https://erwannlc.github.io/mhtml-wizard/), then you can create the issue here: https://github.com/gildas-lormeau/mhtml-to-html/issues

**Comment by nuthrash** [comment]
Hi @Marvin-Brouwer
 
> Minor comments I have: I see "No file" for a couple of seconds when opening an mhtml file. I imagine that's because it's still being generated. And, obsidian can't find the file yet. Perhaps you can start with a default HTML page that shows some kind of loading text?

It is a good idea to show a temporary converting information page while invoking the mhtml-to-html library, and I will add this feature to next version of this plugin. 

> Additionally, I have a couple of very big mhtml files which seem to take quite a while to reopen. I'm not sure if it's because of rendering or the on the fly HTML conversion. But perhaps some caching mechanism may help here?

I think you can try to convert some very big mhtml files to html files by the online converting tool which gildas-lormeau mentioned, then open the converted html files by this plugin to see if it also take quite a while to reopen.
If the situation is the same. I guess the caching mechanism is useless, because these files are too big to open and render (in fact, the opened content is already in the memory, system should reuse it on reopening procedure).
If opening these converted html files is quick. I think the best way is to convert these mhtml files to html files, then put them back to the vault, instead of expecting converting them on the fly or the caching mechanism, especially when you often open these files in a high frequency.

**Comment by Marvin-Brouwer** [comment]
@nuthrash I had a look at mhtml-wizard thing.
And I do see a significant difference.
If you'd like I can share both the mhtml and the html file with you so you can conduct some of your own tests?

**Comment by nuthrash** [comment]
@Marvin-Brouwer 
> If you'd like I can share both the mhtml and the html file with you so you can conduct some of your own tests?

Yes, I need your mhtml and html files as samples for my testing.

**Comment by Marvin-Brouwer** [comment]
@nuthrash here are the files: https://we.tl/t-CrfreYPW4z the link is valid for 3 days.
Please don't mind the contents, I was harassing ChatGPT with a brainfart for JavaScript which ended up a failure.
It did create a large mthml file though, so it helps in a way :P

**Comment by Marvin-Brouwer** [comment]
@nuthrash Did my files help you?
Do you need any help testing

**Comment by nuthrash** [comment]
> [@nuthrash](https://github.com/nuthrash) Did my files help you? Do you need any help testing

@Marvin-Brouwer Yes, they are very helpful. I have tested them and compare with my other big mhtml files.
I have tried to open some around 20MB mhtml files, they are so big with lots pictures, and the Obsidian open them always faster than your files.
According the testing result, I realize the content structure of a big mhtml file is the key point of file opening time issue. In other words, more simple structures and smaller sizes would result in shorter file opening and converting time.
By the way, current converting flow would consume lots memory.

I will try to modify the converting flow to release unused memory.
And I am evaluating some cache mechanisms to see if any of them can improve loading time for reopened files.
